client/src/utils/CountdownTimer.tsx

```diff
@@ -7,49 +8,45 @@ interface TimeLeft {
   seconds: number
 }
 
-export function CountdownTimer(props: { deadline: Date }) {
-  function calculateTimeLeft(): TimeLeft {
-    const difference: number = props.deadline.getTime() - Date.now()
-
-    if (difference > 0) {
-      return {
-        days: Math.floor(difference / (1000 * 60 * 60 * 24)),
-        hours: Math.floor((difference / (1000 * 60 * 60)) % 24),
-        minutes: Math.floor((difference / 1000 / 60) % 60),
-        seconds: Math.floor((difference / 1000) % 60),
-      }
-    }
+function calculateTimeLeft(deadline: Date): TimeLeft {
+  const difference: number = deadline.getTime() - Date.now()
+
+  if (difference > 0) {
     return {
-      days: 0,
-      hours: 0,
-      minutes: 0,
-      seconds: 0,
+      days: Math.floor(difference / (1000 * 60 * 60 * 24)),
+      hours: Math.floor((difference / (1000 * 60 * 60)) % 24),
+      minutes: Math.floor((difference / 1000 / 60) % 60),
+      seconds: Math.floor((difference / 1000) % 60),
     }
   }
+  return {
+    days: 0,
+    hours: 0,
+    minutes: 0,
+    seconds: 0,
+  }
+}
+
+export function isInPast(date: Date): boolean {
+  const timeLeft = calculateTimeLeft(date)
+  return !timeLeft.days && !timeLeft.hours && !timeLeft.minutes && !timeLeft.seconds
+}
 
-  const [timeLeft, setTimeLeft] = useState<TimeLeft>(calculateTimeLeft())
+export function CountdownTimer(props: { deadline: Date; timeOutMessage: string }) {
+  const [timeLeft, setTimeLeft] = useState<TimeLeft>(calculateTimeLeft(props.deadline))
 
   useEffect(() => {
     const timeout = setTimeout(() => {
-      setTimeLeft(calculateTimeLeft())
+      setTimeLeft(calculateTimeLeft(props.deadline))
     }, 1000)
     return () => clearTimeout(timeout)
   })
 
-  const timerComponents = []
-
-  if (timeLeft.days) {
-    timerComponents.push(<span>{timeLeft.days} days, </span>)
-  }
-  if (timeLeft.hours) {
-    timerComponents.push(<span>{timeLeft.hours} hours, </span>)
-  }
-  if (timeLeft.days) {
-    timerComponents.push(<span>{timeLeft.minutes} minutes, </span>)
-  }
-  if (timeLeft.days) {
-    timerComponents.push(<span>{timeLeft.seconds} seconds</span>)
-  }
+  const timeUp = isInPast(props.deadline)
 
-  return <span>{timerComponents.length ? timerComponents : <span>Time's up!</span>}</span>
+  return !timeUp ? (
+    <span>{String.Format('in {days}:{hours:00}:{minutes:00}:{seconds:00}', timeLeft)}</span>
```

Comment:
I think it would be better if we said the start date in a human way. Probably using [`date-fns.formatRelative`](https://date-fns.org/v2.10.0/docs/formatRelative)

I see you added a library called `@date-io/date-fns` that wraps around `date-fns`, so that probably exposes the `formatRelative` in some way.

If that works well, we could remove the `typescript-string-operations` to reduce the project dependencies

Reply:
We'll probably have to work on the Timer again at some point, I'll do that then :) I left a TODO